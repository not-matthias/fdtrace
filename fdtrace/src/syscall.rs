use std::ops::{Deref, DerefMut};

#[allow(non_camel_case_types)]
pub type pid_t = i32;
#[allow(non_camel_case_types)]
pub type tid_t = i32;

#[allow(non_camel_case_types)]
pub type fd_t = u64;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Syscall {
    pub ts: u64,
    pub pid: pid_t,
    pub tid: tid_t,
    pub raw: RawSyscall,
}

impl Syscall {
    pub fn from_parts(data: &str) -> Option<Self> {
        let parts = data.split(";");
        Self::from_parts_iter(parts)
    }

    pub fn from_parts_iter<'a>(mut parts: impl Iterator<Item = &'a str>) -> Option<Self> {
        Some(Self {
            ts: parts.next()?.parse().ok()?,
            pid: parts.next()?.parse().ok()?,
            tid: parts.next()?.parse().ok()?,
            raw: RawSyscall::from_parts_iter(parts)?,
        })
    }
}

impl Deref for Syscall {
    type Target = RawSyscall;

    fn deref(&self) -> &Self::Target {
        &self.raw
    }
}

impl DerefMut for Syscall {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.raw
    }
}

/// # Covered syscalls
///
/// - File creation and opening: open, openat.
/// - File descriptor operations: close, read, write.
/// - File removal: unlink, unlinkat, rmdir.
/// - File renaming: rename, renameat.
/// - Directory creation: mkdir, mkdirat.
#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(serde::Serialize))]
#[rustfmt::skip]
pub enum RawSyscall {
    Execve { path: String },

    Open { path: String, flags: u64, mode: u64 },
    OpenExit { ret: i64 },

    OpenAt { dirfd: fd_t, path: String, flags: u64 },
    OpenAtExit { ret: i64 },

    Close { fd: fd_t },
    CloseExit { ret: i64 },

    Read { fd: fd_t, count: usize },
    ReadExit { read: i64 },

    Write { fd: fd_t, count: usize },
    WriteExit { written: i64 },
}

impl RawSyscall {
    pub fn from_parts(data: &str) -> Option<Self> {
        let parts = data.split(";");
        Self::from_parts_iter(parts)
    }

    /// Parses the ';' separated syscall (e.g. `read;42;42`)
    pub fn from_parts_iter<'a>(mut parts: impl Iterator<Item = &'a str>) -> Option<Self> {
        macro_rules! parse_syscall {
            ($syscall:ident, $($field:ident),*) => {
                Some(RawSyscall::$syscall {
                    $($field: parts.next()?.parse().ok()?,)*
                })
            };
        }

        match parts.next()? {
            "execve" => parse_syscall!(Execve, path),

            "open" => parse_syscall!(Open, path, flags, mode),
            "open_exit" => parse_syscall!(OpenExit, ret),

            "openat" => parse_syscall!(OpenAt, dirfd, path, flags),
            "openat_exit" => parse_syscall!(OpenAtExit, ret),

            "close" => parse_syscall!(Close, fd),
            "close_exit" => parse_syscall!(CloseExit, ret),

            "read" => parse_syscall!(Read, fd, count),
            "read_exit" => parse_syscall!(ReadExit, read),

            "write" => parse_syscall!(Write, fd, count),
            "write_exit" => parse_syscall!(WriteExit, written),

            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tracer::BpfTracer;

    #[test]
    fn test_parse_threaded() {
        let raw_trace = include_str!("../data/threaded.txt");
        let syscalls = BpfTracer::parse_trace(&raw_trace).unwrap();

        insta::assert_json_snapshot!(syscalls);
    }

    #[test]
    fn test_parse_multisession() {
        let raw_trace = include_str!("../data/multisession.txt");
        let syscalls = BpfTracer::parse_trace(&raw_trace).unwrap();

        insta::assert_json_snapshot!(syscalls);
    }

    #[test]
    fn test_parse_read_exit() {
        let parts = "read_exit;832";
        let syscall = RawSyscall::from_parts(parts).unwrap();
        assert_eq!(syscall, RawSyscall::ReadExit { read: 832 });
    }

    #[test]
    fn test_parse_close() {
        let parts =
            "20429708185183;105898;105898;openat;18446631905284431216;;18446631905284431216";
        let syscall = Syscall::from_parts(parts).unwrap();
        assert_eq!(syscall.ts, 18836222727359);
        assert_eq!(syscall.pid, 99220);
        assert_eq!(syscall.tid, 99220);
        assert_eq!(
            syscall.raw,
            RawSyscall::Close {
                fd: 18446631905284438752
            }
        );
    }
}
