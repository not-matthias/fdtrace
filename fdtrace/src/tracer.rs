use crate::syscall::Syscall;
use std::path::Path;
use tempfile::NamedTempFile;

pub struct BpfTracer {
    syscalls: Vec<Syscall>,
}

impl BpfTracer {
    pub fn trace(program: &Path) -> anyhow::Result<Self> {
        let tmpfile = NamedTempFile::new()?;
        let cmd = std::process::Command::new("bpftrace")
            .arg("-c")
            .arg(program)
            .arg("-o")
            .arg(tmpfile.path())
            // TODO: Fix this path
            .arg("scripts/fdtrace.bt")
            .output()?;

        if !cmd.status.success() {
            let error = String::from_utf8_lossy(&cmd.stderr);
            return Err(anyhow::anyhow!("{error}"));
        }

        let output = std::fs::read_to_string(tmpfile)?;
        Self::parse_trace(&output)
    }

    pub fn parse_trace(trace: &str) -> anyhow::Result<Self> {
        let parse_pid = |line: &str| {
            line.split(';')
                .next()
                .and_then(|p: &str| p.parse::<u32>().ok())
                .unwrap_or_default()
        };

        let mut target_pid = None;
        let mut syscalls = Vec::new();
        for line in trace.lines() {
            let Some(target_pid) = target_pid else {
                // The output contains many other processes logs as well, which is not what we
                // want. We need to find the 'execve' syscall to find the process id of our
                // target process.
                //
                if line.contains("execve") {
                    target_pid = Some(parse_pid(line));
                }

                continue;
            };

            // After we have our target_pid, we can filter out all the other logs that
            // aren't related to this process.
            //
            if parse_pid(line) != target_pid {
                continue;
            }

            let parts = line.split(';').skip(1).collect::<Vec<_>>().join(";");

            if let Some(syscall) = Syscall::from_parts(&parts) {
                syscalls.push(syscall);
            } else {
                // TODO: Return as Result<_, _>
                log::error!("Failed to parse syscall: {}", parts);
            }
        }

        Ok(Self { syscalls })
    }

    /// Prints the syscalls to a file
    pub fn print_to_file(&self, file: &str) {
        use std::io::Write;

        let mut file = std::fs::File::create(file).unwrap();
        for syscall in &self.syscalls {
            writeln!(file, "{:?}", syscall).unwrap();
        }
    }

    /// Prints the syscalls to stdout
    pub fn debug_print(&self) {
        for syscall in &self.syscalls {
            println!("{:?}", syscall);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Run with: sudo cargo t
    #[test]
    fn test_trace_ls() {
        let tracer =
            BpfTracer::trace("/nix/store/sbxy42ph4gjlg567vaz1kihmgiwqa5dh-system-path/bin/ls")
                .unwrap();
        assert_ne!(tracer.syscalls.len(), 0);
    }
}
