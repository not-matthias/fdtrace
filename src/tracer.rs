use crate::syscall::{RawSyscall, Syscall};
use std::{io::Write, path::Path};
use tempfile::NamedTempFile;

pub struct BpfTracer {
    syscalls: Vec<Syscall>,
}

impl BpfTracer {
    pub fn trace(program: &Path) -> anyhow::Result<Self> {
        let script = {
            let mut file = NamedTempFile::new()?;
            writeln!(file, "{}", include_str!("../data/fdtrace.bt"))?;
            file
        };

        let tmpfile = NamedTempFile::new()?;
        let cmd = std::process::Command::new("bpftrace")
            .arg("-c")
            .arg(program)
            .arg("-o")
            .arg(tmpfile.path())
            .arg(script.path())
            .output()?;

        if !cmd.status.success() {
            let error = String::from_utf8_lossy(&cmd.stderr);
            return Err(anyhow::anyhow!("{error}"));
        }

        let output = std::fs::read_to_string(tmpfile)?;
        std::fs::write("raw_trace.txt", &output).unwrap();

        Ok(Self {
            syscalls: Self::parse_trace(&output)?,
        })
    }

    pub fn parse_trace(trace: &str) -> anyhow::Result<Vec<Syscall>> {
        let mut target_pid = None;

        let mut syscalls = Vec::new();
        for line in trace.lines().skip(1) {
            if line.starts_with("Lost ") {
                log::warn!("Lost events: {line}");
                continue;
            }

            let syscall =
                Syscall::from_parts(line).unwrap_or_else(|| panic!("Failed to parse: {line}"));

            // The output contains many other processes logs as well, which is not what we
            // want. We need to find the 'execve' syscall to find the process id of our
            // target process.
            //
            let Some(target_pid) = target_pid else {
                if let RawSyscall::Execve { path } = &syscall.raw {
                    log::info!("Target process: {:?}", path);
                    target_pid = Some(syscall.pid);
                }
                continue;
            };

            // After we have our target_pid, we can filter out all the other logs that
            // aren't related to this process.
            //
            if syscall.pid != target_pid {
                continue;
            }

            syscalls.push(syscall);
        }

        Ok(syscalls)
    }

    pub fn syscalls(&self) -> &[Syscall] {
        &self.syscalls
    }

    pub fn take_syscalls(self) -> Vec<Syscall> {
        self.syscalls
    }

    /// Prints the syscalls to a file
    pub fn print_to_file(&self, file: &str) {
        use std::io::Write;

        let mut file = std::fs::File::create(file).unwrap();
        for syscall in &self.syscalls {
            writeln!(file, "{:?}", syscall).unwrap();
        }
    }

    /// Prints the syscalls to stdout
    pub fn debug_print(&self) {
        for syscall in &self.syscalls {
            println!("{:?}", syscall);
        }
    }
}
