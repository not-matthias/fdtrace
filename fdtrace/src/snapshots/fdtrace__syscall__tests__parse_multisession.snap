---
source: fdtrace/src/syscall.rs
expression: syscalls
---
[
  {
    "ts": 21197750001946,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750006966,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750008406,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750009190,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750009856,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750010780,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750012417,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750013725,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750014453,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750015072,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750015686,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750016552,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750017374,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/c1pfdwy1xgb80s7z66w3l7a7nkdvmfnw-rust-mixed/lib/rust..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750019524,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750020211,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/c1pfdwy1xgb80s7z66w3l7a7nkdvmfnw-rust-mixed/lib/rust..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750020778,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750021337,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/c1pfdwy1xgb80s7z66w3l7a7nkdvmfnw-rust-mixed/lib/rust..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750022224,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750036318,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/outpu..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750037293,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750037996,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/outpu..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750038662,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750039289,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/outpu..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750039968,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750040812,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/etc/ld..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750042024,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750042365,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/gl..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750043230,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750043851,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/gl..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750044382,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750044923,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/li..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750045646,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750046383,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/rxganm4ibf31qngal3j3psp20mak37yy-xgcc-13.2.0-libgcc/..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750048112,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750048713,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/rxganm4ibf31qngal3j3psp20mak37yy-xgcc-13.2.0-libgcc/..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750049227,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750049773,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/rxganm4ibf31qngal3j3psp20mak37yy-xgcc-13.2.0-libgcc/..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750052431,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": 4
      }
    }
  },
  {
    "ts": 21197750052776,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 832
      }
    }
  },
  {
    "ts": 21197750054156,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 832
      }
    }
  },
  {
    "ts": 21197750073346,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Close": {
        "fd": 4
      }
    }
  },
  {
    "ts": 21197750075228,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "CloseExit": {
        "ret": 0
      }
    }
  },
  {
    "ts": 21197750076604,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750077832,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750078093,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/home/not-matthias/Documents/technical/git/syscall-tracer/targe..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750079003,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750079238,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/c1pfdwy1xgb80s7z66w3l7a7nkdvmfnw-rust-mixed/lib/rust..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750080146,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": -2
      }
    }
  },
  {
    "ts": 21197750080494,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/li..",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750081680,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": 4
      }
    }
  },
  {
    "ts": 21197750081926,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 832
      }
    }
  },
  {
    "ts": 21197750082594,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 832
      }
    }
  },
  {
    "ts": 21197750106083,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Close": {
        "fd": 4
      }
    }
  },
  {
    "ts": 21197750107048,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "CloseExit": {
        "ret": 0
      }
    }
  },
  {
    "ts": 21197750247373,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/proc/self/maps",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750262159,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": 4
      }
    }
  },
  {
    "ts": 21197750267478,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 1024
      }
    }
  },
  {
    "ts": 21197750274761,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 1024
      }
    }
  },
  {
    "ts": 21197750282905,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 1024
      }
    }
  },
  {
    "ts": 21197750286355,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 1024
      }
    }
  },
  {
    "ts": 21197750288772,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 1024
      }
    }
  },
  {
    "ts": 21197750291308,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 1024
      }
    }
  },
  {
    "ts": 21197750293104,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 1024
      }
    }
  },
  {
    "ts": 21197750294269,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 417
      }
    }
  },
  {
    "ts": 21197750295559,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Close": {
        "fd": 4
      }
    }
  },
  {
    "ts": 21197750296892,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "CloseExit": {
        "ret": 0
      }
    }
  },
  {
    "ts": 21197750307894,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/etc/hosts",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750312848,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": 4
      }
    }
  },
  {
    "ts": 21197750314022,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/etc/passwd",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750315283,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": 5
      }
    }
  },
  {
    "ts": 21197750315606,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAt": {
        "dirfd": 4294967196,
        "path": "/etc/hostname",
        "flags": 524288
      }
    }
  },
  {
    "ts": 21197750318115,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "OpenAtExit": {
        "ret": 6
      }
    }
  },
  {
    "ts": 21197750321291,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 62
      }
    }
  },
  {
    "ts": 21197750322687,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 62
      }
    }
  },
  {
    "ts": 21197750322906,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 4,
        "count": 32
      }
    }
  },
  {
    "ts": 21197750323095,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 0
      }
    }
  },
  {
    "ts": 21197750324994,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Write": {
        "fd": 1,
        "count": 62
      }
    }
  },
  {
    "ts": 21197750328294,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "WriteExit": {
        "written": 62
      }
    }
  },
  {
    "ts": 21197750329414,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Write": {
        "fd": 1,
        "count": 1
      }
    }
  },
  {
    "ts": 21197750329807,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "WriteExit": {
        "written": 1
      }
    }
  },
  {
    "ts": 21197750331770,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 5,
        "count": 4089
      }
    }
  },
  {
    "ts": 21197750332500,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 4089
      }
    }
  },
  {
    "ts": 21197750332681,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 5,
        "count": 32
      }
    }
  },
  {
    "ts": 21197750332894,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 0
      }
    }
  },
  {
    "ts": 21197750333338,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Write": {
        "fd": 1,
        "count": 4151
      }
    }
  },
  {
    "ts": 21197750333812,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "WriteExit": {
        "written": 4151
      }
    }
  },
  {
    "ts": 21197750334016,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Write": {
        "fd": 1,
        "count": 1
      }
    }
  },
  {
    "ts": 21197750334730,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "WriteExit": {
        "written": 1
      }
    }
  },
  {
    "ts": 21197750336537,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 6,
        "count": 4151
      }
    }
  },
  {
    "ts": 21197750337055,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 7
      }
    }
  },
  {
    "ts": 21197750337229,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Read": {
        "fd": 6,
        "count": 4144
      }
    }
  },
  {
    "ts": 21197750337424,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "ReadExit": {
        "read": 0
      }
    }
  },
  {
    "ts": 21197750337683,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Write": {
        "fd": 1,
        "count": 4158
      }
    }
  },
  {
    "ts": 21197750343984,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "WriteExit": {
        "written": 4158
      }
    }
  },
  {
    "ts": 21197750344266,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Write": {
        "fd": 1,
        "count": 1
      }
    }
  },
  {
    "ts": 21197750344572,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "WriteExit": {
        "written": 1
      }
    }
  },
  {
    "ts": 21198750532596,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Close": {
        "fd": 6
      }
    }
  },
  {
    "ts": 21198750536908,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "CloseExit": {
        "ret": 0
      }
    }
  },
  {
    "ts": 21198750537187,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Close": {
        "fd": 5
      }
    }
  },
  {
    "ts": 21198750537700,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "CloseExit": {
        "ret": 0
      }
    }
  },
  {
    "ts": 21198750537823,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "Close": {
        "fd": 4
      }
    }
  },
  {
    "ts": 21198750538272,
    "pid": 110089,
    "tid": 110089,
    "raw": {
      "CloseExit": {
        "ret": 0
      }
    }
  }
]
